experimental/rocm/tracing.c

```diff
@@ -128,6 +157,23 @@ iree_status_t iree_hal_rocm_tracing_context_allocate(
         timestamp_period);
   }
 
+  context->smutrace_enabled = false;
+  void (*dl) = dlopen(SMUTRACE_DLL, RTLD_LAZY);
+  if (dl) {
+    context->f_smuDumpInit = (SmuDumpInitFunc) dlsym(dl, "smuDumpInit");
+    context->f_smuDumpEnd = (SmuDumpEndFunc) dlsym(dl, "smuDumpEnd");
+    context->f_smuDumpOnce = (SmuDumpOnceFunc) dlsym(dl, "smuDumpOnce");
+    context->f_regDumpOnce = (RegDumpOnceFunc) dlsym(dl, "regDumpOnce");
+    context->f_sviDumpOnce = (SviDumpOnceFunc) dlsym(dl, "sviDumpOnce");
+    context->f_smuGetTraceRate = (SmuGetTraceRate) dlsym(dl, "getSmuVariablesCaptureRate");
+    context->f_regGetTraceRate = (RegGetTraceRate) dlsym(dl, "getRegisterExpressionCaptureRate");
+    context->smutrace_enabled = (context->f_smuDumpInit && context->f_smuDumpEnd && context->f_smuDumpOnce &&
```

Comment:
move this  init to its own function that returns an iree_status_t - then you can just bail if anything is not loaded instead of leaving things in an inconsistent state

```c
static iree_status_t iree_hal_rocm_tracing_initialize_smutrace(...) {
  // load library with iree_dynamic_library_load_from_files
  // get symbols with iree_dynamic_library_lookup_symbol and RETURN_IF_ERROR
}
```
then here just
```c
  iree_status_t smutrace_status = iree_hal_rocm_tracing_initialize_smutrace(...);
  if (iree_status_is_ok(status)) {
    context->smutrace.enabled = true;
  } else {
    context->smutrace.enabled = false;
    IREE_TRACE_MESSAGE(...);
    iree_status_ignore(smutrace_status);
  }
```

(getting the status allows for logging it via IREE_TRACE_MESSAGE_DYNAMIC by converting it to a string, but that can come later)